### PR TITLE
Bug 1144132 - Activity opener visibility state

### PR DESCRIPTION
**Pull request by alivedise**

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3c1212c7caeca67eae729adf9eedfc51222ee957

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=97a10b42e16e0493b11263d01b7cd07f1279d8e7

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f6da0de93e54221ed4e685eb8208535cef236c61

File: apps/system/test/unit/child_window_factory_test.js [hunk 1]
```diff
@@ -362,6 +362,24 @@ suite('system/ChildWindowFactory', function() {
...
+    app1.element.dispatchEvent(new CustomEvent('_launchactivity',
+      {
+        detail: fakeActivityDetail
+      }));
+    this.sinon.stub(app1, 'setVisible');
+    spy.getCall(0).returnValue.element
+        .dispatchEvent(new CustomEvent('_opened', {
+          detail: spy.getCall(0).returnValue
+        }));
+    assert.isTrue(app1.setVisible.calledWith(false));
```
Comment (etiennesegonzac) [inline apps/system/test/unit/child_window_factory_test.js]: shouldn't we assert that the we call it with `false, true` ?

File: apps/system/test/unit/child_window_factory_test.js [hunk 2]
```diff
@@ -362,6 +362,24 @@ suite('system/ChildWindowFactory', function() {
...
+    var spy = this.sinon.spy(window, 'ActivityWindow');
+    app1.cwf = new ChildWindowFactory(app1);
+    this.sinon.stub(app1, 'isActive').returns(true);
+    this.sinon.stub(app1, 'isVisible').returns(true);
+    app1.element.dispatchEvent(new CustomEvent('_launchactivity',
+      {
+        detail: fakeActivityDetail
+      }));
+    this.sinon.stub(app1, 'setVisible');
+    spy.getCall(0).returnValue.element
```
Comment (etiennesegonzac) [inline apps/system/test/unit/child_window_factory_test.js]: let's extract `spy.getCall(0).returnValue` into an `activity` variable

File: apps/system/test/unit/child_window_factory_test.js [hunk 3]
```diff
@@ -370,18 +388,19 @@ suite('system/ChildWindowFactory', function() {
...
     var stubSetOrientation = this.sinon.stub(app1, 'setOrientation');
-    var stubRequestForeground = this.sinon.stub(app1, 'requestForeground');
+    this.sinon.stub(app1, 'setVisible');
     spy.getCall(0).returnValue.element
         .dispatchEvent(new CustomEvent('_closing', {
           detail: spy.getCall(0).returnValue
         }));
     assert.isTrue(stubSetOrientation.called);
-    assert.isTrue(stubRequestForeground.called);
+    assert.isTrue(app1.setVisible.calledWith(true));
```
Comment (etiennesegonzac) [inline apps/system/test/unit/child_window_factory_test.js]: we should assert `true, true`

File: apps/system/test/unit/app_window_manager_test.js [hunk 4]
```diff
@@ -545,6 +555,17 @@ suite('system/AppWindowManager', function() {
 
     });
 
+    test('Launch activity', function() {
```
Comment (etiennesegonzac) [inline apps/system/test/unit/app_window_manager_test.js]: `Launch activity dispatched on top most window`

File: apps/system/test/unit/app_window_test.js [hunk 5]
```diff
@@ -1499,6 +1499,7 @@ suite('system/AppWindow', function() {
 
       var stubApp2SetVisible = this.sinon.stub(app2, 'setVisible');
       app1.frontWindow = app2;
+      this.sinon.stub(app2, 'isActive').returns(true);
```
Comment (etiennesegonzac) [inline apps/system/test/unit/app_window_test.js]: we also need to add a small test covering that `doNotPropagate` is respected

File: apps/system/test/unit/child_window_factory_test.js (hunk 1 above)
Comment (alivedise) [inline apps/system/test/unit/child_window_factory_test.js]: done

File: apps/system/test/unit/child_window_factory_test.js (hunk 2 above)
Comment (alivedise) [inline apps/system/test/unit/child_window_factory_test.js]: done

File: apps/system/test/unit/child_window_factory_test.js (hunk 3 above)
Comment (alivedise) [inline apps/system/test/unit/child_window_factory_test.js]: done

File: apps/system/test/unit/app_window_manager_test.js (hunk 4 above)
Comment (alivedise) [inline apps/system/test/unit/app_window_manager_test.js]: done

File: apps/system/test/unit/app_window_test.js (hunk 5 above)
Comment (alivedise) [inline apps/system/test/unit/app_window_test.js]: done

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f9d95f8f2d72d7f674c334cce367450b5a33d734

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5f11d3257b86aecc0a567ed61b1e06718ed5e9ce

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bb54ada70b2e0f897dc664075c020fdf9a8d0d42

**Pull request merged**